## Assignment #1 DATA 730

This assignment tests that you have the course tech stack set up. If you do, you will be able to complete this notebook and turn-in your work by pushing to your assignment repository.

## Tech stack checklist

*Fill out the checklist below once you have completed each task:*

- [ ] GitHub account
- [ ] SageMaker Studio Lab account

**Create the `data730` conda environment by right clicking the `environment.yml` file from the repo in SageMaker Studio Lab and selecting `Build conda Environment`.** 

In [1]:
# uncomment the following line and execute this cell to demonstrate you have created the data730 conda environment
system('conda env list | grep data730', intern = TRUE)

[1] "data730              * C:\\Users\\baile\\miniconda3\\envs\\data730"        
[2] "data730-lecture        C:\\Users\\baile\\miniconda3\\envs\\data730-lecture"

<p style='color:blue'>I still have not been able to register a SageMaker account so I set this environment up on my local Anaconda installation on my Windows laptop. I changed the JupyterLab terminal to use Git Bash instead of PowerShell as the default so the <span style='background-color:lightgray; color:black'>&ensp;grep&ensp;</span> command above would not error. Also I'm running these commands in the <b>data730</b> environment that I built using the environment.yml in the unit-1-assignment-CBailey589 repository.</p>

In [2]:
# uncomment the following line and execute this cell to demonstrate that you have installed the tidymodels package
library('tidymodels')

── Attaching packages ────────────────────────────────────────────────────────────────────────────── tidymodels 1.4.1 ──

✔ broom        1.0.9     ✔ recipes      1.3.1
✔ dials        1.4.2     ✔ rsample      1.3.1
✔ dplyr        1.1.4     ✔ tailor       0.1.0
✔ ggplot2      3.5.2     ✔ tidyr        1.3.1
✔ infer        1.0.9     ✔ tune         2.0.0
✔ modeldata    1.5.1     ✔ workflows    1.3.0
✔ parsnip      1.3.3     ✔ workflowsets 1.1.1
✔ purrr        1.1.0     ✔ yardstick    1.3.2

── Conflicts ───────────────────────────────────────────────────────────────────────────────── tidymodels_conflicts() ──
✖ purrr::discard() masks scales::discard()
✖ dplyr::filter()  masks stats::filter()
✖ dplyr::lag()     masks stats::lag()
✖ recipes::step()  masks stats::step()



## Let's do some statistical inference!

Read [this article](https://www.theguardian.com/world/2002/jan/04/euro.eu2). 

Execute the code below to investigate whether the coin is fair...

In [3]:
# These functions simulate tossing a fair coin n_tosses number of times for n_iterations number of iterations

toss_fair_coin_n_times = function(n_tosses = 250) {
    result = ifelse(runif(n = n_tosses) < 0.5, 'heads', 'tails')
    result
}

repeat_the_experiment = function(n_iterations, n_tosses = 250) {
    n_heads_each_iteration = c()
    for (i in 1:n_iterations) {
        n_heads = sum(toss_fair_coin_n_times(n_tosses) == 'heads')
        n_heads_each_iteration = c(n_heads_each_iteration, n_heads)
    } 
    n_heads_each_iteration
}

In [8]:
# Let's do 10,000 experiments of 250 tosses and see how often our outcome deviates from what was actually observed (139 heads)

n_iterations = 10000
n_tosses = 250
observed_deviation = 139 - 125 # 139 heads observed, expected value for a fair coin of 125

n_heads_each_experiment = repeat_the_experiment(n_iterations, n_tosses)

x = mean(abs(n_heads_each_experiment - 125) >= observed_deviation)
paste0('With a fair coin we would expect a deviation as large or larger than what we observed ', x*100, '% of times')

[1] "With a fair coin we would expect a deviation as large or larger than what we observed 8.6% of times"

**Do you think the coin was fair?** (Answer below in a markdown cell.)

Just keeping track of the results I get to see how consistent they are:
1) 8.38%
2) 8.35%
3) 8.69%
4) 9.14%
5) 8.60%

By this test we're saying that approximately 8-9% of the time we would expect a coin that we know to be fair to to have produced results as 'lopsided' as the Belgium coin produced. I consider this output from our test to be interpretable like a p-value. If we assumed that the standard p-value for coin flip fairness determination were 0.05, then it would need to be less than 5% for us to say with confidence that the coin was unfair. At 8-9% I think it's likely that the Belgium coin was fair and it was just coincidence that both the students and the writers happened to come up with 139 heads vs 111 tails. 

In [10]:
# Let's do 10,000 experiments of 250 tosses and see how often our outcome deviates from what was actually observed (139 heads)
observed_deviation = 138 - 125 # 139 heads observed, expected value for a fair coin of 125

n_heads_each_experiment = repeat_the_experiment(n_iterations, n_tosses)

x = mean(abs(n_heads_each_experiment - 125) >= observed_deviation)
paste0('With a fair coin we would expect a deviation as large or larger than what we observed ', x*100, '% of times')

[1] "With a fair coin we would expect a deviation as large or larger than what we observed 11.29% of times"

In [11]:
# Let's do 10,000 experiments of 250 tosses and see how often our outcome deviates from what was actually observed (139 heads)
observed_deviation = 137 - 125 # 139 heads observed, expected value for a fair coin of 125

n_heads_each_experiment = repeat_the_experiment(n_iterations, n_tosses)

x = mean(abs(n_heads_each_experiment - 125) >= observed_deviation)
paste0('With a fair coin we would expect a deviation as large or larger than what we observed ', x*100, '% of times')

[1] "With a fair coin we would expect a deviation as large or larger than what we observed 14.45% of times"

In [12]:
# Let's do 10,000 experiments of 250 tosses and see how often our outcome deviates from what was actually observed (139 heads)
observed_deviation = 135 - 125 # 139 heads observed, expected value for a fair coin of 125

n_heads_each_experiment = repeat_the_experiment(n_iterations, n_tosses)

x = mean(abs(n_heads_each_experiment - 125) >= observed_deviation)
paste0('With a fair coin we would expect a deviation as large or larger than what we observed ', x*100, '% of times')

[1] "With a fair coin we would expect a deviation as large or larger than what we observed 22.72% of times"

In [13]:
# Let's do 10,000 experiments of 250 tosses and see how often our outcome deviates from what was actually observed (139 heads)
observed_deviation = 126 - 125 # 139 heads observed, expected value for a fair coin of 125

n_heads_each_experiment = repeat_the_experiment(n_iterations, n_tosses)

x = mean(abs(n_heads_each_experiment - 125) >= observed_deviation)
paste0('With a fair coin we would expect a deviation as large or larger than what we observed ', x*100, '% of times')

[1] "With a fair coin we would expect a deviation as large or larger than what we observed 95.01% of times"

This is very unscientific I'm sure, but I just wanted to see how the results would change if I changed the observed_deviation value. Based purely on feel after looking at those results, I think the Belgium coin was probably fair. 